sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -0,0 +1,200 @@
+   * This can be called only after prepareForUpdates() has been called in the current thread.
+   */
+  def commit(): Long
+
+  /** Cancel all the updates that have been made to the store. */
+  def cancel(): Unit
+
+  /**
+   * Iterator of store data after a set of updates have been committed.
+   * This can be called only after commitUpdates() has been called in the current thread.
+   */
+  def iterator(): Iterator[InternalRow]
+
+  /**
+   * Iterator of the updates that have been committed.
+   * This can be called only after commitUpdates() has been called in the current thread.
+   */
+  def updates(): Iterator[StoreUpdate]
+
+  /**
+   * Whether all updates have been committed
+   */
+  def hasCommitted: Boolean
+}
+
+
+trait StateStoreProvider {
+
+  /** Get the store with the existing version. */
+  def getStore(version: Long): StateStore
+
+  /** Optional method for providers to allow for background management */
+  def manage(): Unit = { }
+}
+
+sealed trait StoreUpdate
+case class ValueAdded(key: InternalRow, value: InternalRow) extends StoreUpdate
+case class ValueUpdated(key: InternalRow, value: InternalRow) extends StoreUpdate
+case class KeyRemoved(key: InternalRow) extends StoreUpdate
+
+
+/**
+ * Companion object to [[StateStore]] that provides helper methods to create and retrive stores
+ * by their unique ids.
+ */
+private[state] object StateStore extends Logging {
+
+  private val MANAGEMENT_TASK_INTERVAL_SECS = 60
+
+  private val loadedProviders = new mutable.HashMap[StateStoreId, StateStoreProvider]()
+  private val managementTimer = new Timer("StateStore Timer", true)
+  @volatile private var managementTask: TimerTask = null
+
+  /** Get or create a store associated with the id. */
+  def get(storeId: StateStoreId, directory: String, version: Long): StateStore = {
+    require(version >= 0)
+    val storeProvider = loadedProviders.synchronized {
+      startIfNeeded()
+      val provider = loadedProviders.getOrElseUpdate(
+          storeId, new HDFSBackedStateStoreProvider(storeId, directory))
+      reportActiveInstance(storeId)
+      provider
+    }
+    storeProvider.getStore(version)
+  }
+
+  def remove(storeId: StateStoreId): Unit = loadedProviders.synchronized {
+    loadedProviders.remove(storeId)
+  }
+
+  def isLoaded(storeId: StateStoreId): Boolean = loadedProviders.synchronized {
+    loadedProviders.contains(storeId)
+  }
+
+  /** Unload and stop all state store provider */
+  def stop(): Unit = loadedProviders.synchronized {
+    loadedProviders.clear()
+    if (managementTask != null) {
+      managementTask.cancel()
+      managementTask = null
+      logInfo("StateStore stopped")
+    }
+  }
+
+  private def startIfNeeded(): Unit = loadedProviders.synchronized {
+    if (managementTask == null) {
+      managementTask = new TimerTask {
+        override def run(): Unit = { manageAll() }
+      }
+      val periodMs = Option(SparkEnv.get).map(_.conf) match {
```

Comment:
SparkEnv.get returns null when a SparkContext is not active. I am afraid that this StateStore object's timer will keep ticking even if there is not SparkContext active (in local mode), and this is to make sure that this background this does not fail unexpectedly. 

Actually on second thought, the timer does not need to be started if SparkEnv.get returns nulls, it only needs to be lazily stopped when SparkEnv.get becomes null (which it already does). I will modify this logic.